subsys/storage/storage_area/storage_area_store.c

```diff
@@ -0,0 +1,1097 @@
+{
+	struct storage_area_store_data *data = store->data;
+	const struct storage_area *area = store->area;
+	const size_t secpos = data->sector * store->sector_size;
+	uint8_t buf[SAS_MAX(SAS_MINBUFSIZE, area->write_size)];
+	struct storage_area_iovec wr = {
+		.data = buf,
+	};
+	int rc = 0;
+
+	memset(buf, SAS_FILLVALUE, sizeof(buf));
+	while (data->loc < store->sector_size) {
+		sa_off_t wroff = secpos + data->loc;
+
+		wr.len = SAS_MIN(sizeof(buf), store->sector_size - data->loc);
+		rc = storage_area_writev(area, wroff, &wr, 1U);
+		if (rc != 0) {
+			break;
+		}
+
+		data->loc += wr.len;
+	}
+
+	if (rc != 0) {
+		LOG_DBG("failed to fill sector %d", data->sector);
+	}
+
+	return rc;
+}
+
+static int store_erase_block(const struct storage_area_store *store)
+{
+	const struct storage_area *area = store->area;
+	const size_t erase_size = area->erase_size;
+	struct storage_area_store_data *data = store->data;
+	int rc = 0;
+
+	if ((data->sector * store->sector_size) % erase_size != 0U) {
+		goto end;
+	}
+
+	size_t sblock = (data->sector * store->sector_size) / erase_size;
+	size_t bcnt = MAX(1U, store->sector_size / erase_size);
+
+	rc = storage_area_erase(area, sblock, bcnt);
+
+	if (rc != 0) {
+		LOG_DBG("erase failed at block %d", sblock);
+	}
+end:
+	return rc;
+}
+
+static int store_advance_scb(const struct storage_area_store *store,
+			     const struct storage_area_store_compact_cb *cb)
+{
+	ARG_UNUSED(cb);
+
+	const struct storage_area *area = store->area;
+	struct storage_area_store_data *data = store->data;
+	int rc = 0;
+
+	if (STORAGE_AREA_FOVRWRITE(area)) {
+		rc = store_fill_sector(store);
+		if (rc != 0) {
+			goto end;
+		}
+	}
+
+	sector_advance(store, &data->sector, 1U);
+	if (data->sector == 0U) {
+		data->wrapcnt++;
+	}
+
+	data->loc = 0U;
+
+	if ((!STORAGE_AREA_FOVRWRITE(area)) &&
+	    (!STORAGE_AREA_AUTOERASE(area))) {
+		rc = store_erase_block(store);
+		if (rc != 0) {
+			goto end;
+		}
+	}
+
+	rc = store_add_cookie(store);
+end:
+	return rc;
+}
+
+static int store_move_record(struct storage_area_record *record,
+			     const struct storage_area_store_compact_cb *cb)
+{
+	const struct storage_area_store *store = record->store;
+	struct storage_area_store_data *data = store->data;
+
+	if ((cb->move == NULL) || (!cb->move(record))) {
+		return 0;
+	}
+
+	if (!store_record_valid(record)) {
+		LOG_DBG("invalid record, skipping move");
+		return 0;
+	}
+
+	const struct storage_area *area = store->area;
+	const size_t sector_size = store->sector_size;
+	const size_t align = area->write_size;
+	const struct storage_area_record dest = {
+		.store = (struct storage_area_store *)store,
+		.sector = data->sector,
+		.loc = data->loc,
+		.size = record->size};
+	const size_t rdpos = record->sector * sector_size + record->loc;
+	const size_t wrpos = data->sector * sector_size + data->loc;
+	const size_t alsize =
+		SAS_ALIGNUP(SAS_HDRSIZE + record->size + SAS_CRCSIZE, align);
+	uint8_t buf[SAS_MAX(SAS_MINBUFSIZE, align)];
```

Comment:
This is actually stack limiting. For devices with small alignment the reads are combined to one larger read, for devices with large alignment the reads are limited to the align size. The aim here is to work on all devices, including these with a write-block-size equal to 512.

Reply:
That is NOT how stack works, and it is not limiting anything. CONFIG_MAIN_STACK_SIZE does limiting for main thread, whatever it is set to, and that is not available for anything else. And that "dynamic" allocation allocates from that limit anyway, by decreasing base pointer by the size used in function frame, and moving all incoming frames more or less down the address space.

Reply:
With "this is actually stack limiting" I intended to say "this is stack usage limiting", it will only use what is required by the alignment, unless the alignment is small in which case it will use the `SAS_MINBUFSIZE`.